_posts/2019-12-02-Nx.md

```diff
@@ -0,0 +1,177 @@
+All your apps and libraries will be part of a dependency graph.
+
+<img src="{{ '/img/2019-12-02-nx/dependency_graph.png' | prepend: site.baseurl }}" alt="Dependency graph"  class="image fit" style="margin:0px auto; max-width: 500px;">
+
+The dependency graph will give you an up-to-date version of all the applications and services and the dependencies between them.
+It can also help you to get a better understanding of the architecture.
+
+## Building
+When new code is introduced or something has been refactored, Nx will be able to find out which dependencies are affected by your changes.
+This gives you the opportunity to only build the affected dependencies and only run the tests of those affected ones.
+For your build pipeline, that means that testing the affected apps can greatly improve the build time when dealing with large workspaces.
+
+For example, if we edit something in the shop app, our dependency graph will look like this:
+
+<img src="{{ '/img/2019-12-02-nx/dependency_graph_affected.png' | prepend: site.baseurl }}" alt="Dependency graph of affected apps" class="image fit" style="margin:0px auto; max-width: 500px;">
+
+Targetting the affected code can be done using the affected command.
+For example, if you want to run the unit tests for the affected code:
+```
+nx affected:test
+```
+Inside of your CI, you would like to compare the changes between your branch and master.
+This can be done by appending the two branches on the previous command:
+```
+nx affected:test --base=origin/master --head=your-branch
+```
+
+## Restrictions
+You may want to categorize libraries in different domains and limit the dependencies between those domains to improve maintainability.
+Nx can help you with this, by restricting access between domains or by allowing the dependency flow to go only one way around.
+To configure and manage these restrictions, you can apply tags in the nx.json file for apps and libraries.
+
+# Full stack applications
+With a monorepo, you can have both your frontend and your backend applications under the same repository.
+This enables you to share code between them.
+With Nx, you can do this by using libraries like mentioned before.
+These libraries will expose the code using a public API.
+
+The code inside these libraries can be imported afterwards in your apps, without having to fetch them from a npm registry.
+Think about how easily you can share an interface between your frontend and backend, without having to break it on one side.
+Whenever something changes in the API, both sides will get updated so they will both remain in sync.
+
+# Conclusion
+Having worked within a monorepo myself for the last months, I really appreciate the way it improved the daily workflow.
+It encourages me to write more reusable code and to keep features small.
+Other teams may have the same requirements, so code can be easily migrated to libs.
```

Comment:
Though beware to not create a giant "shared" lib. Any change to that will affect a lot of applications. It's better to have more small libs than less fat libs.